lib/maths/CDataFrameCategoryEncoder.cc

```diff
@@ -429,24 +435,18 @@ std::uint64_t CDataFrameCategoryEncoder::COneHotEncoding::checksum() const {
     return CChecksum::calculate(seed, m_HotCategory);
 }
 
-void CDataFrameCategoryEncoder::COneHotEncoding::acceptPersistInserter(
+void CDataFrameCategoryEncoder::COneHotEncoding::acceptPersistInserterSubroutine(
     core::CStatePersistInserter& inserter) const {
-    CEncoding::acceptPersistInserter(inserter);
     core::CPersistUtils::persist(ONE_HOT_ENCODING_CATEGORY_TAG, m_HotCategory, inserter);
 }
 
-bool CDataFrameCategoryEncoder::COneHotEncoding::acceptRestoreTraverser(core::CStateRestoreTraverser& traverser) {
-    do {
-        const std::string& name = traverser.name();
-        RESTORE(ENCODING_INPUT_COLUMN_INDEX_TAG,
-                core::CPersistUtils::restore(ENCODING_INPUT_COLUMN_INDEX_TAG,
-                                             m_InputColumnIndex, traverser))
-        RESTORE(ENCODING_MIC_TAG,
-                core::CPersistUtils::restore(ENCODING_MIC_TAG, m_Mic, traverser))
-        RESTORE(ONE_HOT_ENCODING_CATEGORY_TAG,
-                core::CPersistUtils::restore(ONE_HOT_ENCODING_CATEGORY_TAG,
-                                             m_HotCategory, traverser))
-    } while (traverser.next());
+bool CDataFrameCategoryEncoder::COneHotEncoding::acceptRestoreTraverserSubroutine(
+    core::CStateRestoreTraverser& traverser) {
+
+    const std::string& name = traverser.name();
+    RESTORE_NO_LOOP(ONE_HOT_ENCODING_CATEGORY_TAG,
+                    core::CPersistUtils::restore(ONE_HOT_ENCODING_CATEGORY_TAG,
+                                                 m_HotCategory, traverser))
```

Comment:
I think missing `traverser.next()` is probably an error here. I'd be inclined to just leave this in the usual for loop and not worry with `RESTORE_NO_LOOP`

Reply:
The advancement is happening in the method `CDataFrameCategoryEncoder::CEncoding::acceptRestoreTraverser(core::CStateRestoreTraverser& traverser)` of the base class. So I don't need to do it here. Otherwise, I would need to traverse the state twice: once to extract the members of the base class and once to extract the members of the derived class. Instead, I implemented the template pattern and restore all members in a single traversing pass.

Reply:
Oh right, I see what you're doing now. What I was envisioning was base class like:
```
do {
    const std::string& name = traverser.name();
    RESTORE(ENCODING_INPUT_COLUMN_INDEX_TAG,
            core::CPersistUtils::restore(ENCODING_INPUT_COLUMN_INDEX_TAG,
                                         m_InputColumnIndex, traverser))
    RESTORE(ENCODING_MIC_TAG,
            core::CPersistUtils::restore(ENCODING_MIC_TAG, m_Mic, traverser))
    break;
} while (traverser.next());
return this->acceptRestoreTraverser(traverser);
```

Reply:
I'm not saying what you've done is worse :)